### PR TITLE
Implement numpy autowrapping in remote context

### LINKED ISSUE #1819: Intercept Numpy automatically in remote context
Right now numpy needs to be manually given as "remote numpy" to be properly used in remote contexts.
We need to intercept it and make it work normally without user interactions as much as possible.

### PR DESCRIPTION
**Pull request by vnlitvinov**
## What do these changes do?
This replaces `numpy` module whenever a user code executes `import numpy` with a special class that returns different things in local and remote contexts.
It also registers special reducers (support for pickling), as, when trying to pickle an object from "local" numpy when "remote" is activated leads to `TypeError: different class` without special reducer.

Note that access to `numpy` is always local for `modin.*` and `pandas.*` packages, as NumPy is used internally there and is not needed to be proxied to the remote.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1819
- [ ] tests added and passing

### CODE REVIEW

File: modin/experimental/pandas/numpy_wrap.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    except ImportError:
+        pass
+    else:
+        _EXCLUDE_MODULES.append(rpyc)
+    _EXCLUDE_PATHS = tuple(
+        os.path.dirname(mod.__file__) + os.sep for mod in _EXCLUDE_MODULES
+    )
+
+    class InterceptedNumpy(types.ModuleType):
+        __own_attrs__ = set(["__own_attrs__"])
```
Comment (anmyachev) [inline modin/experimental/pandas/numpy_wrap.py:41]: please add docstring

File: modin/experimental/pandas/numpy_wrap.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+        def __update_engine(self, _):
+            if execution_engine.get() == "Cloudray":
+                from modin.experimental.cloud import get_connection
+
+                self.__swap_numpy(get_connection().modules["numpy"])
+            else:
+                self.__swap_numpy()
+
+        def __make_reducer(self, name):
```
Comment (anmyachev) [inline modin/experimental/pandas/numpy_wrap.py:75]: please add docstring

File: modin/experimental/pandas/numpy_wrap.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+        def __get_numpy(self):
+            frame = sys._getframe()
+            try:
+                caller_file = frame.f_back.f_back.f_code.co_filename
+            except AttributeError:
+                return self.__current_numpy
+            finally:
+                del frame
+            if any(caller_file.startswith(p) for p in _EXCLUDE_PATHS):
```
Comment (anmyachev) [inline modin/experimental/pandas/numpy_wrap.py:112]:
```suggestion
            if any(caller_file.startswith(path) for path in _EXCLUDE_PATHS):
```

File: modin/experimental/pandas/numpy_wrap.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+                    # TODO: see if replacing all "real numpy" things in reduced[1:] is needed
+                    return (reconstruct,) + reduced[1:]
+
+                self.__reducers[name] = reducer
+            return reducer
+
+        def __get_numpy(self):
+            frame = sys._getframe()
+            try:
+                caller_file = frame.f_back.f_back.f_code.co_filename
```
Comment (anmyachev) [inline modin/experimental/pandas/numpy_wrap.py:107]: need comment with explanation

File: modin/experimental/pandas/numpy_wrap.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+        def __setattr__(self, name, value):
+            if name in self.__own_attrs__:
+                super().__setattr__(name, value)
+            else:
+                setattr(self.__get_numpy(), name, value)
+
+        def __delattr__(self, name):
+            if name not in self.__own_attrs__:
+                delattr(self.__get_numpy(), name)
```
Comment (anmyachev) [inline modin/experimental/pandas/numpy_wrap.py:130]: Please add docstring for these methods. It's not obvious, that in `__setattr__` you use `super()`, but don't do this in `__getattr__` and `__getattr__`.

File: modin/experimental/pandas/__init__.py [hunk 6]
```diff
@@ -14,6 +14,8 @@
 import os
 
 os.environ["MODIN_EXPERIMENTAL"] = "True"
+
+from .numpy_wrap import _CAUGHT_NUMPY  # noqa F401
```
Comment (anmyachev) [inline modin/experimental/pandas/__init__.py:18]: why `_CAUGHT_NUMPY` in `__init__`?

File: modin/experimental/pandas/numpy_wrap.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+import sys
+
+_CAUGHT_NUMPY = "numpy" not in sys.modules
+try:
```
Comment (devin-petersohn) [inline modin/experimental/pandas/numpy_wrap.py:17]: A short comment about why this is needed will help.

File: modin/experimental/pandas/numpy_wrap.py (hunk 1 above)
Comment (vnlitvinov) [reply modin/experimental/pandas/numpy_wrap.py:41]: added

File: modin/experimental/pandas/numpy_wrap.py (hunk 2 above)
Comment (vnlitvinov) [reply modin/experimental/pandas/numpy_wrap.py:75]: added

File: modin/experimental/pandas/numpy_wrap.py (hunk 3 above)
Comment (vnlitvinov) [reply modin/experimental/pandas/numpy_wrap.py:112]: renamed to `mod_path`

File: modin/experimental/pandas/numpy_wrap.py (hunk 4 above)
Comment (vnlitvinov) [reply modin/experimental/pandas/numpy_wrap.py:107]: added

File: modin/experimental/pandas/numpy_wrap.py (hunk 5 above)
Comment (vnlitvinov) [reply modin/experimental/pandas/numpy_wrap.py:130]: added

File: modin/experimental/pandas/__init__.py (hunk 6 above)
Comment (vnlitvinov) [reply modin/experimental/pandas/__init__.py:18]: added

File: modin/experimental/pandas/numpy_wrap.py (hunk 7 above)
Comment (vnlitvinov) [reply modin/experimental/pandas/numpy_wrap.py:22]: comment added

**Comment by anmyachev** [comment]
@vnlitvinov LGTM! Let's wait CI result.

**Pull request merged**